### ISSUE TITLE
Cargo vendor with patch section and local sources does not vendor the local files

### ISSUE THREAD

**Issue by Firstyear** [rust-lang/cargo#9172]
**Problem**
If a patch section contains a specification with a "path" override, cargo vendor does not vendor these content to the vendor directory. 

**Steps**
Add a patch statement such as:

```
[patch.crates-io]
mycrate = { path = "../mycrate" }
```

Then run `cargo vendor`

**Possible Solution(s)**
Cargo vendor should copy the content of the crate (excluding git artefacts) to the vendor directory. 

**Notes**

cargo 1.50.0 (f04e7fab7 2021-02-04)

**Comment by ianrrees** [comment]
I think this is a more general problem actually, it seems `cargo vendor` doesn't vendor crates referred to by paths.

**Comment by Firstyear** [comment]
Paths can be a tricky one as the source could be in the same workspace though. I think it would only be a path that refers to a crate outside the workspace that would be needed here?

**Comment by ianrrees** [comment]
Yeah, and I guess the implementation would only copy files that would be left after a `cargo clean`.

Mainly, my comment was just meant to say that the patch statement in your original example may be expected to fail.  I found this issue after being surprised that `cargo vendor` didn't copy a dependency that's referred to by path.  It would be nice if there were at least an option to also vendor such dependencies.

**Comment by stevenroose** [comment]
I proposed https://github.com/rust-lang/cargo/pull/12858 to solve this issue.

**Comment by weihanglo** [comment]
I am not going to propose `cargo-vendor` to support vendoring all path dependencies. But for this `[patch]` use case specifically, should we reconsider to lift the limitation a bit?

[`[patch]` in config.toml](https://doc.rust-lang.org/cargo/reference/unstable.html#patch-in-config) is a thing since 1.56. This feature opens a door for people to have a central place to configure their patches. In an enterprise environment people might have a pre-configured `.cargo/config.toml` for their own internal patched dependencies. These patched dependencies might also be pre-downloaded somewhere in the local file system as target patched path dependencies. These all work fine until `cargo vendor` fails.

`[patch]` in config could be a system-wide setting that a user might not have control over, so it is better that `cargo vendor` includes those pointing to local path dependencies.

(This also applies to vendoring path dependencies in general. However, path dependencies are not system-wide configuration hence I am not going to justify that use case.)

**Comment by weihanglo** [comment]
Proposed another proof-of-concept here in <https://github.com/rust-lang/cargo/pull/13347>.

**Comment by Firstyear** [comment]
Sorry to be a pain, this has come up again. We're looking at ways to allow patching of dependencies by maintainers for opensuse, and so having the ability to include path based dependencies would really help as then we can vendor these and commit them for future auditing without needing to resort to a really messy diff based patching process. In the meantime we are able to proceed because we can vendor in git based dependencies that contain fixes we may need, but it's a bit more overhead by comparison.
